lib/services/s3.js

```diff
@@ -338,22 +431,172 @@ AWS.util.update(AWS.S3.prototype, {
+    var bucket = req.params.Bucket || null;
+    var bucketRegionCache = req.service.bucketRegionCache;
+    if (region && bucket && region !== bucketRegionCache[bucket]) {
+      bucketRegionCache[bucket] = region;
+    }
+
+    var cachedRegion;
     if (codes[code] && body.length === 0) {
+      if (bucket && !region) {
+        cachedRegion = bucketRegionCache[bucket] || null;
+        if (cachedRegion !== req.httpRequest.region) {
+          region = cachedRegion;
+        }
+      }
       resp.error = AWS.util.error(new Error(), {
-        code: codes[resp.httpResponse.statusCode],
-        message: null
+        code: codes[code],
+        message: null,
+        region: region
       });
     } else {
       var data = new AWS.XML.Parser().parse(body.toString());
+
+      if (data.Region && !region) {
+        region = data.Region;
+        if (bucket && region !== bucketRegionCache[bucket]) {
+          bucketRegionCache[bucket] = region;
+        }
+      } else if (bucket && !region && !data.Region) {
+        cachedRegion = bucketRegionCache[bucket] || null;
+        if (cachedRegion !== req.httpRequest.region) {
+          region = cachedRegion;
+        }
+      }
+
       resp.error = AWS.util.error(new Error(), {
         code: data.Code || code,
         message: data.Message || null,
-        region: data.Region || null
+        region: region
       });
     }
-    resp.request.service.extractRequestIds(resp);
+    req.service.extractRequestIds(resp);
+  },
+
+  /**
+   * If region was not obtained synchronously, then send async request
+   * to get bucket region for errors resulting from wrong region.
+   *
+   * @api private
+   */
+   requestBucketRegion: function requestBucketRegion(resp, done) {
+    var error = resp.error;
+    var req = resp.request;
+    var bucket = req.params.Bucket || null;
+
+    if (!error || !bucket || error.region || req.operation === 'listObjects' ||
+        (AWS.util.isNode() && req.operation === 'headBucket') ||
+        (error.statusCode === 400 && req.operation !== 'headObject') ||
+        regionRedirectErrorCodes.indexOf(error.code) === -1) {
+      return done();
+    }
+
+    var regionReq = req.service.listObjects({Bucket: bucket, MaxKeys: 0});
+    regionReq._requestRegionForBucket = bucket;
+    regionReq.send(function() {
+      var region = req.service.bucketRegionCache[bucket] || null;
+      error.region = region;
+      done();
+    });
+   },
+
+   /**
+   * For browser only. If NetworkingError received, will attempt to obtain
+   * the bucket region.
+   *
+   * @api private
+   */
+   reqRegionForNetworkingError: function reqRegionForNetworkingError(resp, done) {
+    if (!AWS.util.isBrowser()) {
+      return done();
+    }
+    var error = resp.error;
+    var request = resp.request;
+    var bucket = request.params.Bucket;
+    if (!error || error.code !== 'NetworkingError' || !bucket ||
+        request.httpRequest.region === 'us-east-1') {
```

Comment:
Because `NetworkingError` can be caused by other reasons, it can occur when the region is set to `us-east-1`, but it won't be because of incorrect region. A `BadRequest` or `AuthorizationHeaderMalformed` would be received in that case instead, and would be handled by the `extractError` listener. If the region is set to `us-east-1`, this listener wouldn't be doing anything helpful for it anyway, since it is just either changing the region to `us-east-1`, or trying a region request in `us-east-1`. This exclusion filter is needed to prevent this listener from recursively sending off an infinite number of requests that cause the next request to be sent. For example, if there is something wrong with the network and we're receiving a `NetworkingError` because of it, then this listener would keep getting triggered because each request it sends off would receive another `NetworkingError`. So instead of allowing that, this listener will check that the region is not `us-east-1`, because otherwise the `NetworkingError` is not due to incorrect region.